### ISSUE TITLE
ranges in output arguments

### ISSUE THREAD

**Issue by cosunae** [GridTools/gridtools#72]
This is to discuss the current syntax of ranges and its impact for output fields.

What does the following mean??

``` C++
        typedef arg_type<0, range<-1, 1, -1, 1> > out;
```

For input fields, 

``` C++
        typedef const arg_type<0, range<-1, 1, -1, 1> > in;
```

this means we are accessing the argument at 

``` C++
in(i-1), in(i+1), in(j-1), in(j+1)
```

But for the use of the output fields, this would be wrong. 
Does that mean we can write at 
out(i+1) and out(i-1) ?

In the case of output arguments it means, well, extend the block (for this ESF) by this range. 
However, if this makes sense for the automatic extension of the computation of temporaries (from previous ESF), output variables are not private, they are in GMEM/DRAM.
So this will create unnecessary redundant computations of the output field and race conditions from different blocks, as multiple blocks are accessing the same addresses for the grid points at the extended halos.

Also, what is the meaning of the output range in the following example?

``` C++
    struct Lap {
        typedef const arg_type<0, range<-1,1,-1,1> > in;
        typedef arg_type<1, range<-1,1,-1,1>> out;
        typedef boost::mpl::vector<in, out> arg_list;

        template <typename Evaluation>
        GT_FUNCTION
        static void Do(Evaluation const & eval, x_interval) {
            eval(out()) = -4. * eval(in()) + eval(in(i-1)) + eval(in(i+1)) + eval(in(j-1))  +eval(in(j+1))  
        }
    };

    struct hori_diff {
        typedef const arg_type<0, range<-1,1,-1,1> > in;
        typedef arg_type<1, range<-1,1,-1,1>> out;
        typedef boost::mpl::vector<in, out> arg_list;

        template <typename Evaluation>
        GT_FUNCTION
        static void Do(Evaluation const & eval, x_interval) {
            eval(out()) = eval(Call<Lap>(Call<Lap>(in) ) )
        }
    }
```

We could want to achieve other purposes with this syntax, namely:
- Extend the range of computation at the global boundaries, not the halos of each block, but really the global border of the whole domain. Well this is not what we use the ranges for. Another syntax for this use case would be to specify that in the make_esf?

``` C++
 gridtools::make_mss
       (
            execute<forward>(),
            gridtools::make_esf<copy_functor, boundary<-1,1,-1,1> >( ),
            gridtools::make_esf<lap_functor, boundary<-1,1,-1,1> > ( )
       )
```
- We could want to shift the computation indices of the iteration space for. But I dont see the use case, but still, in that case, what we need there is to shift, not to control the range (i.e. grow or shrink)

``` C++
    struct Lap {
        typedef const arg_type<0, range<-1,1,-1,1> > in;
        typedef arg_type<1, shift<-1,1>> out;
        typedef boost::mpl::vector<in, out> arg_list;

        template <typename Evaluation>
        GT_FUNCTION
        static void Do(Evaluation const & eval, x_interval) {
            eval(out()) = -4. * eval(in()) + eval(in(i-1)) + eval(in(i+1)) + eval(in(j-1))  +eval(in(j+1))  
        }
    };
```

In summary, the use of 

``` C++
            gridtools::make_esf<copy_functor, boundary<-1,1,-1,1> >( ),
```

and shifts (in case this is really a use case, which I am not convinced)

``` C++
        typedef arg_type<1, shift<-1,1>> out;
```

Would cover the use of ranges we do today. And it avoids the miss-use of the ranges for output arguments I pointed out before. We could then enforce to not have ranges for output args
